[storage-file-share] README files for samples directories

willmtemple opened the pull request:
Addresses #5679 for file-shares.

I'm starting here and getting some feedback before I proceed to the rest of the packages. I'll want to add similar files for all the packages that are GA or that are going GA. I'll leave some comments on the files below with some open questions for anyone with input.

We need to have README files in each of the packages' `samples/typescript` and `samples/javascript` folders (with YAML front-matter configured like this) in order for our sample code to be ingested to docs.microsoft.com/samples

This is, to the best of my knowledge, the recommended formatting.

CC @kaerm @ramya-rao-a

willmtemple commented on sdk/storage/storage-file-share/samples/javascript/README.md:
```diff
@@ -0,0 +1,68 @@
+products:
+- azure
+- azure-storage
+urlFragment: storage-file-share-javascript
```
@kaerm my understanding is that this `urlFragment` directive here will make it so the sample URL ends up being pinned to docs.microsoft.com/samples/Azure/azure-sdk-for-js/storage-file-share-javascript , right? Is there a preferred naming scheme for this?

kaerm replied:
that's how I understand it as well

XiaoningLiu replied:
Is there any preview link for the generted docs?

kaerm replied:
i do not think so. Once we have it merged we can see how it generates, and if change is needed it will be automatically picked up

willmtemple commented on sdk/storage/storage-file-share/samples/javascript/README.md:
````diff
@@ -0,0 +1,68 @@
+```bash
+npm install @azure/storage-file-share
+```
+2. Copy and modify the sample code, replacing the relative `require('../../')` statements with `require('@azure/storage-file-share')`)
````
It's really unfortunate that one of the steps has to be "edit this code", since we are using relative imports in these files, which won't work for the samples after they are zipped up. The only alternative to editing the files is to clone the whole repo and build the client sdk using Rush, which seems even more unfortunate for the end-user of the samples.

ramya-rao-a replied:
Ideally, we shouldn't have this step. I'd rather have `require('@azure/storage-file-share')` checked in to git and we change it to `require('../..')` before trying to run it.

Something similar is what I tried for service bus at https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/package.json#L44

willmtemple replied:
I like the idea of pre-processing them fro CI and checking in the camera-ready copies. I would want to move the script to a shared location, though. I think we probably also want package.json (and tsconfig for the typescript ones) for the samples, that way the user experience on docs.microsoft.com is just download/extract the zip, run `npm install` (`npm run build` as well for ts, and we don't have to instruct the user to globally install TypeScript), then run a sample in Node. It would be awesome for these to be as turnkey as possible.

It feels to me like a lot of this process could be automated using a script to generate such files from a template. If we put good doc comments at the top of every sample explaining them, then we can use that text to generate the table of files in the readme and we could generate the title from the package name, the one-line description from package metadata, etc.

XiaoningLiu replied:
Vote for the idea : )

ramya-rao-a replied:
+1 for the automation, lets log an issue to track that.
In this PR, lets do the what's needed to have the samples light up in the docs

willmtemple commented:
Looks like these READMEs would need to be excluded from the README checker (or have their own checker) in Azure Pipelines.

XiaoningLiu commented on sdk/storage/storage-file-share/samples/javascript/README.md:
```diff
@@ -0,0 +1,68 @@
+
+The sample code is compatible with Node.js. The samples are compatible with Node.js >= 8.0.0, except the samples that use the async `for await` syntax. Samples that use the async `for await` syntax require a version of Node that is compatible with that syntax (>= 10.0.0).
+
+You need [an Azure subscription][freesub] and [an Azure Storage account][azstorage] to run these sample programs. Samples retrieve credentials to access the storage account from environment variables. See each individual sample for details on which environment variables it requires to function.
```
Need to fillin actual links here.

xirzec commented on .docsettings.yml:
```diff
@@ -10,6 +10,8 @@ omitted_paths:
   - "sdk/cosmosdb/cosmos/samples/*"
   - eng/tools/versioning/*
   - samples/Bundling/**
+  - "sdk/**/samples/javascript/README.md"
```
`sdk/**/samples/**/README.md` to collapse it to one line maybe?

xirzec commented on sdk/storage/storage-file-share/samples/javascript/README.md:
```diff
@@ -0,0 +1,68 @@
+
+|__File Name__|__Description__|
+|-------------|---------------|
+|[basic.js][basic]|authenticate with the service using an account name & key (or anonymously with a SAS URL), upload a file, list files and directories, and download a file to a string|
```
English nit: I feel like the descriptions should be full sentences, starting with a capital letter and ending with a period

willmtemple replied:
I originally wrote them as sentences and found that they all began with "This sample" and then had all the current words except in a different tense. So, I decided to change them to make them feel less repetitive when scanning top-to-bottom. If you feel strongly that they should be full sentences, then I can try to write better/more diversified sentences, or I can change them back, but I think it subtracts from the information in the table.

xirzec commented on sdk/storage/storage-file-share/samples/javascript/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## Prerequisites
+
+The sample code is compatible with Node.js. The samples are compatible with Node.js >= 8.0.0, except the samples that use the async `for await` syntax. Samples that use the async `for await` syntax require a version of Node that is compatible with that syntax (>= 10.0.0).
```
```suggestion
Samples are compatible with Node.js >= 8.0.0, except for samples which use `for await` syntax, which require Node >= 10.0.0.
```

xirzec commented on sdk/storage/storage-file-share/samples/typescript/README.md:
```diff
@@ -0,0 +1,79 @@
+
+|__File Name__|__Description__|
+|-------------|---------------|
+|[basic.ts][basic]|authenticate with the service using an account name & key (or anonymously with a SAS URL), upload a file, list files and directories, and download a file to a string|
```
same English nit as JS

xirzec commented on sdk/storage/storage-file-share/samples/typescript/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## Prerequisites
+
+The sample code is compatible with Node.js. The samples are compatible with Node.js >= 8.0.0, except the samples that use the async `for await` syntax. Samples that use the async `for await` syntax require a version of Node that is compatible with that syntax (>= 10.0.0).
```
```suggestion
Samples are compatible with Node.js >= 8.0.0, except for samples which use `for await` syntax, which require Node >= 10.0.0.
```

xirzec commented on sdk/storage/storage-file-share/samples/typescript/README.md:
````diff
@@ -0,0 +1,79 @@
+```bash
+npm install @azure/storage-file-share
+```
+2. Copy and modify the sample code, replacing the relative `import { ... } from '../../src'` statements with `import { ... } from '@azure/storage-file-share'`
````
```suggestion
2. Copy and modify the sample code, replacing the relative `import { ... } from "../../src"` statements with `import { ... } from "@azure/storage-file-share"`
```

willmtemple commented:
In the last set of commits:

- Added a script `common/scripts/prep-samples.js` that will make sample code ready for CI, and added it to the build:ts-samples command.
- Added `tsconfig.json` to ts samples and `package.json` to both samples so that when customers download the samples they will have a ready-to-run directory
- Changed readme instructions so that users don't have to edit any files anymore

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+const args = process.argv.slice(2);
+
+let baseDir;
+if (args.length > 0) {
```
could make this simply `if (args.length)`

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+  console.log("Package name", package.name);
+  fs.readdir(tsDir, { withFileTypes: true }, function(err, files) {
+    if (err) {
+      console.error("Failed to read TypeScript samples directory:", args[0]);
```
why not use `baseDir` here instead of `args[0]`?

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+ * Breadth-first search for files ending in .ts
+ */
+function findAllTsFiles(filesAndFolders) {
+  const q = filesAndFolders.map(function(f) {
```
nit: prefer arrow functions for callbacks

xirzec replied:
e.g. 

```javascript
const q = filesAndFolders.map(f => [f, tsDir]);
````

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+
+    if (file.isDirectory()) {
+      // Enqueue children of this directory to the bfs
+      fs.readdirSync(fullPath, { withFileTypes: true }).forEach(function(
```
nit: arrow function

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+      ) {
+        q.push([child, fullPath]);
+      });
+    } else if (file.isFile() && file.name.match(/.*\.ts$/)) {
```
use `String.endsWith` instead of `match` e.g. `file.name.endsWith(".ts")`

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+
+  let idx = 0;
+
+  while (idx !== q.length) {
```
this feels a little odd to me. How about `while (q.length)` and use `q.shift()` to take from the front of the array?

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+    `Preparing samples for package: ${package.name}@${package.version}`
+  );
+  console.log("Package name", package.name);
+  fs.readdir(tsDir, { withFileTypes: true }, function(err, files) {
```
maybe `util.promisify` and use async/await?

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+ * @param {string} pkgName name of the package to use when looking for package-local imports
+ */
+function enableLocalRun(file, baseDir, pkgName) {
+  const fileContents = fs.readFileSync(file, { encoding: "utf-8" });
```
if you decide to use async/await, might as well make this all async

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+
+  const relativeDir = path.dirname(file.replace(baseDir, ""));
+
+  // `string.length - string.split(path.sep).join("").length` is a dirty but well-supported way to
```
this is pretty clever!

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,108 @@
+      process.exit(1);
+    }
+
+    findAllTsFiles(files).forEach(function(fileName) {
```
nit: use `for/of` instead of `.forEach`

xirzec commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,136 @@
+ * @param {string} tsDir The root of the sample tree to search
+ * @returns
+ */
+async function* findAllTsFiles(tsDir) {
```
generators! my only weakness! ;)

xirzec commented on sdk/storage/storage-file-share/samples/typescript/tsconfig.json:
```diff
@@ -0,0 +1,13 @@
+
+    "outDir": "dist"
+  },
+  "include": ["**/*.ts"]
```
I don't think this is necessary, the default is to spider for all TS files starting with the workspace root, right?

jeremymeng commented on common/scripts/prep-samples.js:
```diff
@@ -0,0 +1,149 @@
+
+  while (idx !== q.length) {
+    // [fs.Dirent, string] (file and dirName part of the full path)
+    const [file, dirName] = q[idx];
```
>file [](start = 11, length = 4)

nit: `entry` a better variable name than `file`?

willmtemple commented:
Merging this. We'll see how these look once docs renders them and then move on from there.
